zcash_client_sqlite/src/wallet/init/migrations/add_transaction_views.rs

```diff
@@ -0,0 +1,356 @@
+        "Add transaction summary views & add fee information to transactions."
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch("ALTER TABLE transactions ADD COLUMN fee INTEGER;")?;
+
+        let mut stmt_list_txs =
+            transaction.prepare("SELECT id_tx, raw, block FROM transactions")?;
+
+        let mut stmt_set_fee =
+            transaction.prepare("UPDATE transactions SET fee = ? WHERE id_tx = ?")?;
+
+        let mut stmt_find_utxo_value = transaction
+            .prepare("SELECT value_zat FROM utxos WHERE prevout_txid = ? AND prevout_idx = ?")?;
+
+        let mut tx_rows = stmt_list_txs.query(NO_PARAMS)?;
+        while let Some(row) = tx_rows.next()? {
+            let id_tx: i64 = row.get(0)?;
+            let tx_bytes: Option<Vec<u8>> = row.get(1)?;
+            let h: u32 = row.get(2)?;
+            let block_height = BlockHeight::from(h);
+
+            // If only transaction metadata has been stored, and not transaction data, the fee
+            // information will eventually be set when the full transaction data is inserted.
+            if let Some(b) = tx_bytes {
+                let tx =
+                    Transaction::read(&b[..], BranchId::for_height(&self.params, block_height))
+                        .map_err(|e| {
+                            WalletMigrationError::CorruptedData(format!(
+                                "Parsing failed for transaction {:?}: {:?}",
+                                id_tx, e
+                            ))
+                        })?;
+
+                let fee_paid = tx.fee_paid(|op| {
+                    stmt_find_utxo_value
+                        .query_row(&[op.hash().to_sql()?, op.n().to_sql()?], |row| {
+                            row.get(0).map(|i| Amount::from_i64(i).unwrap())
+                        })
+                        .map_err(WalletMigrationError::DbError)
+                })?;
+
+                stmt_set_fee.execute(&[i64::from(fee_paid), id_tx])?;
+            }
+        }
+
+        transaction.execute_batch(
+            "CREATE VIEW v_tx_sent AS
+            SELECT transactions.id_tx         AS id_tx,
+                   transactions.block         AS mined_height,
+                   transactions.tx_index      AS tx_index,
+                   transactions.txid          AS txid,
+                   transactions.expiry_height AS expiry_height,
+                   transactions.raw           AS raw,
+                   SUM(sent_notes.value)      AS sent_total,
+                   COUNT(sent_notes.id_note)  AS sent_note_count,
+                   SUM(
+                       CASE
+                           WHEN sent_notes.memo IS NULL THEN 0
+                           WHEN SUBSTR(sent_notes.memo, 0, 2) = X'F6' THEN 0
```

Comment:
This is not technically correct. Per [ZIP 302](https://zips.z.cash/zip-0302#specification):
> - the byte `0xF6` followed by 511 `0x00` bytes, indicating "no memo";
>
> ...
>
> If the memo matches any of these patterns, then this memo is from the future, because these ranges are reserved for future updates to this specification:
> - The first byte has a value of `0xF6`, and the remaining 511 bytes are not all `0x00`.

The entire memo needs to be checked to verify it is an empty memo. (Just being able to check the first byte was [my rationale for keeping the original definition of empty memos from the initial draft](https://github.com/zcash/zips/pull/105#discussion_r597257764), but it was decided to go with the protocol spec definition instead for various reasons.)

That being said, a memo of `null` has previously meant "we don't know what the memo is yet", so a rationale we could use here is that we are just not counting memos that we don't know how to represent to the user. That rationale would technically imply excluding all "future" memos as well, and that feels weird because it's likely that future memo types are meaningful. But if what the SDK is going to use this count for is "how many memos are we going to show the user" then there's no need to count future memo types, in which case we can leave this check as-is (and we should add additional exclusions). Adding exclusions into the view does mean that adding a new future memo type will require migrating the view.

Reply:
And now that I look at this again, how does this work? `memo` is a `BLOB`, and `SUBSTR` is [defined as](https://www.sqlite.org/lang_corefunc.html#substr) (emphasis mine):

> The substr(X,Y,Z) function returns a substring of input string X that begins with the Y-th character and which is Z characters long. If Z is omitted then substr(X,Y) returns all characters through the end of the string X beginning with the Y-th. **The left-most character of X is number 1.** If Y is negative then the first character of the substring is found by counting from the right rather than the left. If Z is negative then the abs(Z) characters preceding the Y-th character are returned. If X is a string then characters indices refer to actual UTF-8 characters. If X is a BLOB then the indices refer to bytes.

Doesn't this mean that `SUBSTR(sent_notes.memo, 0, 2)` is going to attempt to return 2 bytes starting from index 0, which is before the start of the `BLOB`? Should this not be `SUBSTR(sent_notes.memo, 1, 1)`?

And additionally, can we not check against the entire memo with `SUBSTR(sent_notes.memo, 1, 512) = X'F6000000....'`?

Reply:
https://www.sqlite.org/lang_corefunc.html#quote suggests that what we maybe need here is `QUOTE(sent_notes.memo) = X'F60000....'`

Reply:
I tried this with `QUOTE` and it still will not match the empty memo literal.